test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchiveUpgrade.java

```diff
@@ -134,11 +164,75 @@ static void assertCreatedJSA(OutputAnalyzer output) {
         output.shouldContain("Dumping shared data to file");
     }
 
-    static void assertJSAVersionMismatch(OutputAnalyzer output) {
-        output.shouldContain("does not match the required version");
-    }
-
     static void assertUsedJSA(OutputAnalyzer output) {
         output.shouldContain("Mapped dynamic region #0");
     }
+
+    // Fetch JDK artifact depending on platform
+    // If the artifact cannot be found, default to the test.boot.jdk property
+    private static String fetchJDK(int version) throws Throwable {
+        int build;
+        String architecture;
+        HashMap<String, Object> jdkArtifactMap = new HashMap<>();
+        jdkArtifactMap.put("server", "jpg");
+        jdkArtifactMap.put("product", "jdk");
+
+        // Select the correct release build number for each version
+        // *UPDATE THIS* after each release
+        switch(version) {
+            case 19:
+                build = 36;
+                break;
+            case 20:
+                build = 29;
+                break;
+            default:
+                throw new RuntimeException("Unsupported JDK version " + version);
+        }
+        jdkArtifactMap.put("version", version);
+        jdkArtifactMap.put("build_number", build);
+
+        // Get correct file name for architecture
+        if (Platform.isX64()) {
+            architecture = "x";
+        } else if (Platform.isAArch64()) {
+            architecture = "aarch";
+        } else {
+            throw new RuntimeException("Unsupported architecture " + Platform.getOsArch());
+        }
+
+        // File name is bundles/<os>-<architecture>64/jdk-<version>_<os>-<architecture>64_bin.<extension>
+        // Ex: bundles/linux-x64/jdk-19_linux-x64_bin.tar.gz
+        if (Platform.isWindows()) {
+            jdkArtifactMap.put("file", "bundles/windows-x64/jdk-" + version + "_windows-x64_bin.zip");
+            return fetchJDK(jdkArtifactMap, version);
+        } else if (Platform.isOSX()) {
+            jdkArtifactMap.put("file", "bundles/macos-" + architecture + "64/jdk-" + version + "_macos-" + architecture + "64_bin.tar.gz");
+            return fetchJDK(jdkArtifactMap, version) +  ".jdk" + FS + "Contents" + FS + "Home";
+        } else if (Platform.isLinux()) {
+            jdkArtifactMap.put("file", "bundles/linux-" + architecture + "64/jdk-" + version + "_linux-" + architecture + "64_bin.tar.gz");
+            return fetchJDK(jdkArtifactMap, version);
+        } else {
+            throw new RuntimeException("Unsupported operating system " + Platform.getOsName());
+        }
+    }
+
+    // Fetch JDK artifact
+    private static String fetchJDK(HashMap<String, Object> jdkArtifactMap, int version) {
+        String path = null;
+        try {
+            path = ArtifactResolver.resolve("jdk", jdkArtifactMap, true) + "/jdk-" + version;
+            System.out.println("Boot JDK path: " + path);
```

Comment:
`String path` should be declared inside this `try` block and its value should be returned here. That way you don't need to read all the catch block below to find out what value is actually returned.